### DISCUSSION TITLE
Rationale behind relation between Topobank and SurfaceTopography objects

### DISCUSSION THREAD

**Discussion by jotelha** [ContactEngineering/topobank#1062] [Development]
Many topobank database models have a pendant in SurfaceTopography, i.e. surface, topographies, ...

topobank's `Topography` for example provides a method `topography()` that returns  a `SurfaceTopography.SurfaceTopography.Topography/UniformLineScan/NonuniformLineScan` object,

https://github.com/ContactEngineering/topobank/blob/develop/topobank/manager/models.py#L557

Should we implement the same for `topobank.manager.models.Surface` -> `SurfaceTopography.SurfaceContainer` (and possibly topobank `SurfaceCollection` -> list of `SurfaceContainer)`, i.e.
`Surface.surface()` and `SurfaceCollection.surfacecollection()` methods?

**Comment by pastewka** [comment]
A couple of comments here:

The `Topography` class of `topobank` has a few functions:

* It is a database abstraction (and therefore some sort of middleware)
* In addition to database abstraction, it abstracts the file location(s) (this raw + nc). It is also a proxy for lazy loading of the file. In particular, the file should *never* be loaded in the webapp, only in the Celery worker because of memory issues. The webapp only ever opens the file to read metadata but never loads the actual topography data. (Indeed, you'll get a warning if you try to load the data in the webapp process.)
* The `topography` method actually reads the whole file

The `Surface` is a Digital Surface Twin, i.e. a number of topographies plus metadata. 

* The backend `SurfaceTopography` code has no provision of storing the metadata. Indeed, the `SurfaceContainer` class of `SurfaceTopography` is essentially just a list of topographies.
* For memory reasons, a `SurfaceContainer` should never ever actually be constructed in `topobank`. Always make sure to read a topography when processing it and not do a batch read and then batch processing, as this will lead to severe memory issues. There is a proxy container class here that circumvents the memory problem: https://github.com/ContactEngineering/topobank/blob/30db73553df411c7ec32b050fa151ccec0b4116a/topobank/analysis/functions.py#L23
* It probably would make sense to have a `container` convenience method in the `Surface` class that return a proxy container.
